dashboard/app/helpers/aichat_sagemaker_helper.rb

```diff
@@ -1,62 +1,75 @@
     Aws::SageMakerRuntime::Client.new
   end
 
-  # The instruction-tuned version of Mistral accepts formatted instructions where conversation roles
-  # must start with a user prompt and alternate between user and assistant.
-  # Mistral-7B-Instruction LLM instruction format doc at https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1.
+  def self.get_instructions(system_prompt, retrieval_contexts)
+    instructions = ""
+    instructions = system_prompt + " " unless system_prompt.empty?
+    instructions << retrieval_contexts.join(" ") if retrieval_contexts
+    instructions
+  end
+
   def self.format_inputs_for_sagemaker_request(aichat_params, stored_messages, new_message)
-    all_messages = [*stored_messages, new_message]
-    inputs = aichat_params[:systemPrompt] + " "
-    inputs += aichat_params[:retrievalContexts].join(" ") if aichat_params[:retrievalContexts]
-    inputs = SENTENCE_BEGIN_TOKEN + wrap_as_instructions(inputs)
-    all_messages.each do |msg|
-      if msg[:role] == USER
-        inputs += wrap_as_instructions(msg[:chatMessageText])
-      elsif msg[:role] == ASSISTANT
-        # Note that each assistant message in the conversation history is followed by
-        # the end-of-sentence token but a begin-of-sentence token is not required.
-        inputs += msg[:chatMessageText] + SENTENCE_END_TOKEN
-      end
-    end
+    selected_model_id = aichat_params[:selectedModelId]
+    # Add system prompt and retrieval contexts if available to inputs as part of instructions that will be sent to model.
+    instructions = get_instructions(aichat_params[:systemPrompt], aichat_params[:retrievalContexts])
+    model_processor = get_model_processor(selected_model_id)
+    inputs = model_processor.format_model_inputs(instructions, new_message, stored_messages)
+    stopping_strings = model_processor.get_stop_strings
 
     {
       inputs: inputs,
       parameters: {
         temperature: aichat_params[:temperature].to_f,
         max_new_tokens: MAX_NEW_TOKENS,
-        top_p: TOP_P
+        top_p: TOP_P,
+        stop: stopping_strings,
       }
     }
   end
 
-  def self.request_sagemaker_chat_completion(input, endpoint_name)
+  def self.get_model_processor(selected_model_id)
+    case selected_model_id      
+    when MODELS[:PIRATE]
+      return PirateProcessor.new
+    when MODELS[:KAREN]
+      return KarenProcessor.new
+    when MODELS[:ARITHMO]
+      return ArithmoProcessor.new
+    else
+      return MistralProcessor.new
+    end    
+  end
+
+  def self.request_sagemaker_chat_completion(input, selected_model_id)
     create_sagemaker_client.invoke_endpoint(
-      endpoint_name: endpoint_name, # required
+      endpoint_name: selected_model_id, # required
       body: input.to_json, # required
       content_type: "application/json"
     )
   end
 
-  def self.get_sagemaker_assistant_response(sagemaker_response)
+  def self.get_sagemaker_assistant_response(sagemaker_response, selected_model_id)
```

Comment:
minor: not your change, so doesn't have to be part of this PR, but any reason `format_inputs_for_sagemaker_request` and `get_sagemaker_assistant_response` have to be two separate functions? could we just have one function that calls the API and formats the response and returns that to the caller? Wondering because that way we also wouldn't need to pass in `selected_model_id` here again.